src/main/scala/rocket/ICache.scala

```diff
@@ -145,57 +288,144 @@ class ICacheModule(outer: ICache) extends LazyModuleImp(outer)
   val io_hartid = outer.hartIdSinkNodeOpt.map(_.bundle)
+  /** @todo tile Memory mapping I/O base address? */
   val io_mmio_address_prefix = outer.mmioAddressPrefixSinkNodeOpt.map(_.bundle)
+  /** register indicates wheather ITIM is enabled. */
   val scratchpadOn = RegInit(false.B)
+  /** a cut point to SRAM, indicates which SRAM will be used as SRAM or Cache. */
   val scratchpadMax = tl_in.map(tl => Reg(UInt(width = log2Ceil(nSets * (nWays - 1)))))
+  /** line is a minimal granularity accessing to SRAM, check if a line is in the scratchpad or not.
+    * Accessing from [[tl_in]]: convert address to line with `untagBits+log2Ceil(nWays)-1, blockOffBits`(slices of address)
+    * {{{
+    * │          tag         │    set    │offset│
+    *                    ├way┘                    → indicate way location
+    *                    │    line       │
+    * }}}
+    * Accessing from [[io]]: normal cache access @todo
+    */
   def lineInScratchpad(line: UInt) = scratchpadMax.map(scratchpadOn && line <= _).getOrElse(false.B)
+
+  /** scratchpad base address, if exist [[ICacheParams.itimAddr]], add [[ReplicatedRegion]] to base.
+    * @todo seem [[io_hartid]] is not connected?
+    *       maybe when implementing itim, LookupByHartId should be changed to [[]]?
+    */
   val scratchpadBase = outer.icacheParams.itimAddr.map { dummy =>
     p(LookupByHartId)(_.icache.flatMap(_.itimAddr.map(_.U)), io_hartid.get) | io_mmio_address_prefix.get
   }
+
+  /** check an address in the scratchpad address range. */
   def addrMaybeInScratchpad(addr: UInt) = scratchpadBase.map(base => addr >= base && addr < base + outer.size).getOrElse(false.B)
+
+  /** check property this address(paddr) exists in scratchpad.
+    * @todo seems duplicated in `addrMaybeInScratchpad(addr)` between `lineInScratchpad(addr(untagBits+log2Ceil(nWays)-1, blockOffBits))`?
+    */
   def addrInScratchpad(addr: UInt) = addrMaybeInScratchpad(addr) && lineInScratchpad(addr(untagBits+log2Ceil(nWays)-1, blockOffBits))
+
+  /** return the way which will be used as scratchpad for accessing address?
+    * {{{
+    * │          tag         │    set    │offset│
+    *                    └way┘
+    * }}}
+    * @param addr address to be found.
+    */
   def scratchpadWay(addr: UInt) = addr.extract(untagBits+log2Ceil(nWays)-1, untagBits)
+
+  /** check if the selected way is legal.
+    * note: the last way should be reserved to ICache.
+    */
   def scratchpadWayValid(way: UInt) = way < nWays - 1
+
+  /** return the cacheline which will be used as scratchpad for accessing address
+    * {{{
+    * │          tag         │    set    │offset│
+    *                    ├way┘                    → indicate way location
+    *                    │    line       │
+    * }}}
+    * @param addr address to be found.
+   *             applied to slave_addr
+    */
   def scratchpadLine(addr: UInt) = addr(untagBits+log2Ceil(nWays)-1, blockOffBits)
+
+  /** scratchpad stage 0 to stage 1 valid. */
```

Comment:
The `stage N to stage N+1 valid` can just be `scratchpad access valid in stage N`